# Experiment 1 ans 3  Hold Out

##  Base Models

In [11]:
import torch
from torch.utils.data import DataLoader, TensorDataset
from torchvision import models
from torchvision.models import resnet18
import torch.nn as nn
from sklearn.metrics import precision_score, classification_report
import timm

def get_nasnetlarge_base_model(num_classes=2):
    model = timm.create_model('nasnetalarge', pretrained=False)
    model.last_linear = nn.Linear(model.num_features, num_classes)
    return model.to(device)

def get_efficientnetb0_base_model(num_classes=2):
    model = timm.create_model('efficientnet_b0', pretrained=False)
    model.classifier = nn.Linear(model.num_features, num_classes)
    return model.to(device)

def get_resnet18_base_model(num_classes=2):
    model = models.resnet18(pretrained=False)
    model.fc = nn.Linear(model.fc.in_features, num_classes)
    return model.to(device)

In [2]:
# Device configuration
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/base_resnet.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

model = get_resnet18_base_model()
model.load_state_dict(torch.load(best_model_path))
model = model.to(device) 
model.eval()

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

# Calculate weighted precision
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

# Print metrics
print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

# Optional: Full classification report
print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=None`.
  warnings.warn(msg)


Accuracy on the holdout dataset: 71.66%
Weighted Precision on the holdout dataset: 0.7190

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.74      0.66      0.70      5718
     Class 1       0.70      0.77      0.73      5703

    accuracy                           0.72     11421
   macro avg       0.72      0.72      0.72     11421
weighted avg       0.72      0.72      0.72     11421



In [3]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/base_nesnet.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

model = get_nasnetlarge_base_model()
model.load_state_dict(torch.load(best_model_path))
model = model.to(device) 
model.eval()

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

# Calculate weighted precision
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

# Print metrics
print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

# Optional: Full classification report
print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

Accuracy on the holdout dataset: 72.31%
Weighted Precision on the holdout dataset: 0.7232

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.73      0.71      0.72      5718
     Class 1       0.72      0.73      0.73      5703

    accuracy                           0.72     11421
   macro avg       0.72      0.72      0.72     11421
weighted avg       0.72      0.72      0.72     11421



In [4]:
# Device configuration
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/base_effcientnet.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

model = get_efficientnetb0_base_model()
model.load_state_dict(torch.load(best_model_path))
model = model.to(device) 
model.eval()

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

# Calculate weighted precision
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

# Print metrics
print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

# Optional: Full classification report
print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

Accuracy on the holdout dataset: 58.67%
Weighted Precision on the holdout dataset: 0.6095

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.66      0.36      0.47      5718
     Class 1       0.56      0.81      0.66      5703

    accuracy                           0.59     11421
   macro avg       0.61      0.59      0.56     11421
weighted avg       0.61      0.59      0.56     11421



## Pretrained Customized

In [3]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/customized_resnet.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

# Function to recreate the model structure
def get_model(num_classes=2):
    model = models.resnet18(pretrained=True)

    # Freeze all layers except the final fully connected layer
    for param in model.parameters():
        param.requires_grad = False

    # Customize the fully connected layer
    model.fc = nn.Sequential(
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  # Set the model to evaluation mode

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

# Calculate weighted precision
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

# Print metrics
print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

# Optional: Full classification report
print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))


/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 57.95%
Weighted Precision on the holdout dataset: 0.5893

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.62      0.42      0.50      5718
     Class 1       0.56      0.74      0.64      5703

    accuracy                           0.58     11421
   macro avg       0.59      0.58      0.57     11421
weighted avg       0.59      0.58      0.57     11421



In [6]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/customized_nesnet2.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_nasnetlarge_model(num_classes=2):
    model = timm.create_model('nasnetalarge', pretrained=True)

    for name, param in model.named_parameters():
        if "cell_17" in name or "cell_18" in name or "cell_19" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    # Custom fully connected layer
    model.last_linear = nn.Sequential(
        nn.Linear(model.num_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, num_classes),
    )

    return model.to(device)


model = get_nasnetlarge_model(num_classes=2)
try:
    model.load_state_dict(torch.load(best_model_path, map_location=device))
except RuntimeError as e:
    print(f"Error loading state dict: {e}")
    raise
model.eval()

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []
all_probs = []  

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        images_batch = images_batch.to(device)
        labels_batch = labels_batch.to(device)

             with torch.cuda.amp.autocast():
            outputs = model(images_batch)
            _, predicted = torch.max(outputs, 1)
            probabilities = torch.softmax(outputs, dim=1)[:, 1]  

        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())
        all_probs.extend(probabilities.cpu().tolist())

accuracy = correct / total * 100
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

Using device: cuda
Accuracy on the holdout dataset: 71.38%
Weighted Precision on the holdout dataset: 0.7196

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.76      0.63      0.69      5718
     Class 1       0.68      0.79      0.73      5703

    accuracy                           0.71     11421
   macro avg       0.72      0.71      0.71     11421
weighted avg       0.72      0.71      0.71     11421



In [13]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/pretrained_efficientnet.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_model_2(num_classes=5):
    model = models.efficientnet_b0(pretrained=True)

    # Freeze all layers except the final fully connected layer
    for param in model.parameters():
        param.requires_grad = False

    # Customize the fully connected layer
    model.classifier = nn.Sequential(
        nn.Linear(model.classifier[1].in_features, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )

    return model

model = get_model_2(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=EfficientNet_B0_Weights.IMAGENET1K_V1`. You can also use `weights=EfficientNet_B0_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 76.11%
Weighted Precision on the holdout dataset: 0.7616

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.75      0.78      0.77      5718
     Class 1       0.77      0.74      0.76      5703

    accuracy                           0.76     11421
   macro avg       0.76      0.76      0.76     11421
weighted avg       0.76      0.76      0.76     11421



## Finetuned Models

### lr=.01 optimizer=adam

#### Resnet

In [5]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_model_adam_lr0.01.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

# Function to recreate the model structure
def get_model(num_classes=2):
    model = models.resnet18(pretrained=True)

    # Freeze all layers except the final fully connected layer
    for param in model.parameters():
        param.requires_grad = False

    # Customize the fully connected layer
    model.fc = nn.Sequential(
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  # Set the model to evaluation mode

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

# Calculate weighted precision
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

# Print metrics
print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

# Optional: Full classification report
print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 64.35%
Weighted Precision on the holdout dataset: 0.6449

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.63      0.69      0.66      5718
     Class 1       0.66      0.59      0.62      5703

    accuracy                           0.64     11421
   macro avg       0.64      0.64      0.64     11421
weighted avg       0.64      0.64      0.64     11421



#### NasNet

In [9]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_nasnet_adam_lr0.01.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_nasnetlarge_model(num_classes=2):
    model = timm.create_model('nasnetalarge', pretrained=True)

    for name, param in model.named_parameters():
        if "cell_17" in name or "cell_18" in name or "cell_19" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    # Custom fully connected layer
    model.last_linear = nn.Sequential(
        nn.Linear(model.num_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, num_classes),
    )

    return model.to(device)


model = get_nasnetlarge_model(num_classes=2)
try:
    model.load_state_dict(torch.load(best_model_path, map_location=device))
except RuntimeError as e:
    print(f"Error loading state dict: {e}")
    raise
model.eval()

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []
all_probs = []  

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        images_batch = images_batch.to(device)
        labels_batch = labels_batch.to(device)

             with torch.cuda.amp.autocast():
            outputs = model(images_batch)
            _, predicted = torch.max(outputs, 1)
            probabilities = torch.softmax(outputs, dim=1)[:, 1]  

        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())
        all_probs.extend(probabilities.cpu().tolist())

accuracy = correct / total * 100
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

Using device: cuda
Accuracy on the holdout dataset: 66.72%
Weighted Precision on the holdout dataset: 0.6905

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.76      0.49      0.60      5718
     Class 1       0.62      0.84      0.72      5703

    accuracy                           0.67     11421
   macro avg       0.69      0.67      0.66     11421
weighted avg       0.69      0.67      0.66     11421



#### EfficentNet

In [3]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_efficientnet_adam_lr0.01.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"


def get_model_2(num_classes=2):
    model = models.efficientnet_b0(pretrained=True)

    for name, param in model.named_parameters():
        if "features.6" in name or "features.7" in name or "features.8" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    model.classifier = nn.Sequential(
        nn.Linear(model.classifier[1].in_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model_2(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=EfficientNet_B0_Weights.IMAGENET1K_V1`. You can also use `weights=EfficientNet_B0_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 64.35%
Weighted Precision on the holdout dataset: 0.6476

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.67      0.56      0.61      5718
     Class 1       0.62      0.73      0.67      5703

    accuracy                           0.64     11421
   macro avg       0.65      0.64      0.64     11421
weighted avg       0.65      0.64      0.64     11421



### lr=.1 optimizer=adam

#### Resnet

In [7]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_model_adam_lr0.1.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

# Function to recreate the model structure
def get_model(num_classes=2):
    model = models.resnet18(pretrained=True)

    # Freeze all layers except the final fully connected layer
    for param in model.parameters():
        param.requires_grad = False

    # Customize the fully connected layer
    model.fc = nn.Sequential(
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())
accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 52.53%
Weighted Precision on the holdout dataset: 0.6014

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.51      0.96      0.67      5718
     Class 1       0.69      0.09      0.16      5703

    accuracy                           0.53     11421
   macro avg       0.60      0.52      0.41     11421
weighted avg       0.60      0.53      0.41     11421



#### Nasnet

In [12]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_nasnet_adam_lr0.1.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_nasnetlarge_model(num_classes=2):
    model = timm.create_model('nasnetalarge', pretrained=True)

    for name, param in model.named_parameters():
        if "cell_17" in name or "cell_18" in name or "cell_19" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    model.last_linear = nn.Sequential(
        nn.Linear(model.num_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, num_classes),
    )

    return model.to(device)


model = get_nasnetlarge_model(num_classes=2)
try:
    model.load_state_dict(torch.load(best_model_path, map_location=device))
except RuntimeError as e:
    print(f"Error loading state dict: {e}")
    raise
model.eval()

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []
all_probs = []  

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        images_batch = images_batch.to(device)
        labels_batch = labels_batch.to(device)

             with torch.cuda.amp.autocast():
            outputs = model(images_batch)
            _, predicted = torch.max(outputs, 1)
            probabilities = torch.softmax(outputs, dim=1)[:, 1]  

        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())
        all_probs.extend(probabilities.cpu().tolist())

accuracy = correct / total * 100
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

Using device: cuda
Accuracy on the holdout dataset: 49.93%
Weighted Precision on the holdout dataset: 0.2493

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.00      0.00      0.00      5718
     Class 1       0.50      1.00      0.67      5703

    accuracy                           0.50     11421
   macro avg       0.25      0.50      0.33     11421
weighted avg       0.25      0.50      0.33     11421



/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(resul

#### Efficenetnet

In [4]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_efficientnet_adam_lr0.1.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"


def get_model_2(num_classes=2):
    model = models.efficientnet_b0(pretrained=True)

    for name, param in model.named_parameters():
        if "features.6" in name or "features.7" in name or "features.8" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    model.classifier = nn.Sequential(
        nn.Linear(model.classifier[1].in_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model_2(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=EfficientNet_B0_Weights.IMAGENET1K_V1`. You can also use `weights=EfficientNet_B0_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 50.07%
Weighted Precision on the holdout dataset: 0.2507

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.50      1.00      0.67      5718
     Class 1       0.00      0.00      0.00      5703

    accuracy                           0.50     11421
   macro avg       0.25      0.50      0.33     11421
weighted avg       0.25      0.50      0.33     11421



/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(resul

### lr=.2 optimizer=adam

#### Resnet

In [9]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_model_adam_lr0.2.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_model(num_classes=2):
    model = models.resnet18(pretrained=True)

    for param in model.parameters():
        param.requires_grad = False

    model.fc = nn.Sequential(
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 53.47%
Weighted Precision on the holdout dataset: 0.5876

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.52      0.92      0.67      5718
     Class 1       0.66      0.14      0.24      5703

    accuracy                           0.53     11421
   macro avg       0.59      0.53      0.45     11421
weighted avg       0.59      0.53      0.45     11421



#### Nasnet

In [15]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_nasnet_adam_lr0.2.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_nasnetlarge_model(num_classes=2):
    model = timm.create_model('nasnetalarge', pretrained=True)

    for name, param in model.named_parameters():
        if "cell_17" in name or "cell_18" in name or "cell_19" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    # Custom fully connected layer
    model.last_linear = nn.Sequential(
        nn.Linear(model.num_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, num_classes),
    )

    return model.to(device)


model = get_nasnetlarge_model(num_classes=2)
try:
    model.load_state_dict(torch.load(best_model_path, map_location=device))
except RuntimeError as e:
    print(f"Error loading state dict: {e}")
    raise
model.eval()

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []
all_probs = []  

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        images_batch = images_batch.to(device)
        labels_batch = labels_batch.to(device)

             with torch.cuda.amp.autocast():
            outputs = model(images_batch)
            _, predicted = torch.max(outputs, 1)
            probabilities = torch.softmax(outputs, dim=1)[:, 1]  

        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())
        all_probs.extend(probabilities.cpu().tolist())

accuracy = correct / total * 100
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

Using device: cuda
Accuracy on the holdout dataset: 49.93%
Weighted Precision on the holdout dataset: 0.2493

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.00      0.00      0.00      5718
     Class 1       0.50      1.00      0.67      5703

    accuracy                           0.50     11421
   macro avg       0.25      0.50      0.33     11421
weighted avg       0.25      0.50      0.33     11421



/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(resul

#### EfficenetNet

In [5]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_efficientnet_adam_lr0.2.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"


def get_model_2(num_classes=2):
    model = models.efficientnet_b0(pretrained=True)

    for name, param in model.named_parameters():
        if "features.6" in name or "features.7" in name or "features.8" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    model.classifier = nn.Sequential(
        nn.Linear(model.classifier[1].in_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model_2(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=EfficientNet_B0_Weights.IMAGENET1K_V1`. You can also use `weights=EfficientNet_B0_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 53.80%
Weighted Precision on the holdout dataset: 0.5397

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.55      0.44      0.49      5718
     Class 1       0.53      0.64      0.58      5703

    accuracy                           0.54     11421
   macro avg       0.54      0.54      0.53     11421
weighted avg       0.54      0.54      0.53     11421



### lr=.01 optimizer=sgd

#### R

In [11]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_model_sgd_lr0.01.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

# Function to recreate the model structure
def get_model(num_classes=2):
    model = models.resnet18(pretrained=True)

    # Freeze all layers except the final fully connected layer
    for param in model.parameters():
        param.requires_grad = False

    # Customize the fully connected layer
    model.fc = nn.Sequential(
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  # Set the model to evaluation mode

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

# Calculate weighted precision
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

# Print metrics
print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

# Optional: Full classification report
print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 73.53%
Weighted Precision on the holdout dataset: 0.7353

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.74      0.73      0.73      5718
     Class 1       0.73      0.74      0.74      5703

    accuracy                           0.74     11421
   macro avg       0.74      0.74      0.74     11421
weighted avg       0.74      0.74      0.74     11421



#### N

In [18]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_nasnet_sgd_lr0.01.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_nasnetlarge_model(num_classes=2):
    model = timm.create_model('nasnetalarge', pretrained=True)

    for name, param in model.named_parameters():
        if "cell_17" in name or "cell_18" in name or "cell_19" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    # Custom fully connected layer
    model.last_linear = nn.Sequential(
        nn.Linear(model.num_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, num_classes),
    )

    return model.to(device)


model = get_nasnetlarge_model(num_classes=2)
try:
    model.load_state_dict(torch.load(best_model_path, map_location=device))
except RuntimeError as e:
    print(f"Error loading state dict: {e}")
    raise
model.eval()

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []
all_probs = []  

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        images_batch = images_batch.to(device)
        labels_batch = labels_batch.to(device)

             with torch.cuda.amp.autocast():
            outputs = model(images_batch)
            _, predicted = torch.max(outputs, 1)
            probabilities = torch.softmax(outputs, dim=1)[:, 1]  

        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())
        all_probs.extend(probabilities.cpu().tolist())

accuracy = correct / total * 100
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

Using device: cuda
Accuracy on the holdout dataset: 69.36%
Weighted Precision on the holdout dataset: 0.7188

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.79      0.53      0.63      5718
     Class 1       0.64      0.86      0.74      5703

    accuracy                           0.69     11421
   macro avg       0.72      0.69      0.68     11421
weighted avg       0.72      0.69      0.68     11421



#### E

In [6]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_efficientnet_sgd_lr0.01.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"


def get_model_2(num_classes=2):
    model = models.efficientnet_b0(pretrained=True)

    for name, param in model.named_parameters():
        if "features.6" in name or "features.7" in name or "features.8" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    model.classifier = nn.Sequential(
        nn.Linear(model.classifier[1].in_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model
    
model = get_model_2(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=EfficientNet_B0_Weights.IMAGENET1K_V1`. You can also use `weights=EfficientNet_B0_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 77.16%
Weighted Precision on the holdout dataset: 0.7723

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.79      0.75      0.77      5718
     Class 1       0.76      0.80      0.78      5703

    accuracy                           0.77     11421
   macro avg       0.77      0.77      0.77     11421
weighted avg       0.77      0.77      0.77     11421



### lr=.1 optimizer=sgd

#### R

In [13]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_model_sgd_lr0.1.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

# Function to recreate the model structure
def get_model(num_classes=2):
    model = models.resnet18(pretrained=True)

    # Freeze all layers except the final fully connected layer
    for param in model.parameters():
        param.requires_grad = False

    # Customize the fully connected layer
    model.fc = nn.Sequential(
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  # Set the model to evaluation mode

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

# Calculate weighted precision
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

# Print metrics
print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

# Optional: Full classification report
print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 70.26%
Weighted Precision on the holdout dataset: 0.7031

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.71      0.68      0.70      5718
     Class 1       0.69      0.73      0.71      5703

    accuracy                           0.70     11421
   macro avg       0.70      0.70      0.70     11421
weighted avg       0.70      0.70      0.70     11421



#### N

In [21]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_nasnet_sgd_lr0.1.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_nasnetlarge_model(num_classes=2):
    model = timm.create_model('nasnetalarge', pretrained=True)

    for name, param in model.named_parameters():
        if "cell_17" in name or "cell_18" in name or "cell_19" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    # Custom fully connected layer
    model.last_linear = nn.Sequential(
        nn.Linear(model.num_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, num_classes),
    )

    return model.to(device)


model = get_nasnetlarge_model(num_classes=2)
try:
    model.load_state_dict(torch.load(best_model_path, map_location=device))
except RuntimeError as e:
    print(f"Error loading state dict: {e}")
    raise
model.eval()

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []
all_probs = []  

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        images_batch = images_batch.to(device)
        labels_batch = labels_batch.to(device)

             with torch.cuda.amp.autocast():
            outputs = model(images_batch)
            _, predicted = torch.max(outputs, 1)
            probabilities = torch.softmax(outputs, dim=1)[:, 1]  

        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())
        all_probs.extend(probabilities.cpu().tolist())

accuracy = correct / total * 100
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

Using device: cuda
Accuracy on the holdout dataset: 70.22%
Weighted Precision on the holdout dataset: 0.7144

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.77      0.58      0.66      5718
     Class 1       0.66      0.82      0.73      5703

    accuracy                           0.70     11421
   macro avg       0.71      0.70      0.70     11421
weighted avg       0.71      0.70      0.70     11421



#### E

In [7]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_efficientnet_sgd_lr0.1.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"


def get_model_2(num_classes=2):
    model = models.efficientnet_b0(pretrained=True)

    for name, param in model.named_parameters():
        if "features.6" in name or "features.7" in name or "features.8" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    model.classifier = nn.Sequential(
        nn.Linear(model.classifier[1].in_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model_2(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=EfficientNet_B0_Weights.IMAGENET1K_V1`. You can also use `weights=EfficientNet_B0_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 72.31%
Weighted Precision on the holdout dataset: 0.7231

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.73      0.72      0.72      5718
     Class 1       0.72      0.73      0.72      5703

    accuracy                           0.72     11421
   macro avg       0.72      0.72      0.72     11421
weighted avg       0.72      0.72      0.72     11421



### lr=.2 optimizer=sgd

#### R

In [15]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_model_sgd_lr0.2.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

# Function to recreate the model structure
def get_model(num_classes=2):
    model = models.resnet18(pretrained=True)

    # Freeze all layers except the final fully connected layer
    for param in model.parameters():
        param.requires_grad = False

    # Customize the fully connected layer
    model.fc = nn.Sequential(
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  # Set the model to evaluation mode

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

# Calculate weighted precision
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

# Print metrics
print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

# Optional: Full classification report
print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet18_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet18_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 66.93%
Weighted Precision on the holdout dataset: 0.6697

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.66      0.70      0.68      5718
     Class 1       0.68      0.64      0.66      5703

    accuracy                           0.67     11421
   macro avg       0.67      0.67      0.67     11421
weighted avg       0.67      0.67      0.67     11421



#### N

In [24]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device}")

# Paths to the saved model and holdout dataset
best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_nasnet_sgd_lr0.2.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_nasnetlarge_model(num_classes=2):
    model = timm.create_model('nasnetalarge', pretrained=True)

    for name, param in model.named_parameters():
        if "cell_17" in name or "cell_18" in name or "cell_19" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    # Custom fully connected layer
    model.last_linear = nn.Sequential(
        nn.Linear(model.num_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, num_classes),
    )

    return model.to(device)


model = get_nasnetlarge_model(num_classes=2)
try:
    model.load_state_dict(torch.load(best_model_path, map_location=device))
except RuntimeError as e:
    print(f"Error loading state dict: {e}")
    raise
model.eval()

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []
all_probs = []  

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        images_batch = images_batch.to(device)
        labels_batch = labels_batch.to(device)

             with torch.cuda.amp.autocast():
            outputs = model(images_batch)
            _, predicted = torch.max(outputs, 1)
            probabilities = torch.softmax(outputs, dim=1)[:, 1]  

        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())
        all_probs.extend(probabilities.cpu().tolist())

accuracy = correct / total * 100
weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

Using device: cuda
Accuracy on the holdout dataset: 68.71%
Weighted Precision on the holdout dataset: 0.7005

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.75      0.56      0.64      5718
     Class 1       0.65      0.82      0.72      5703

    accuracy                           0.69     11421
   macro avg       0.70      0.69      0.68     11421
weighted avg       0.70      0.69      0.68     11421



#### E

In [8]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_efficientnet_sgd_lr0.2.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"


def get_model_2(num_classes=2):
    model = models.efficientnet_b0(pretrained=True)

    for name, param in model.named_parameters():
        if "features.6" in name or "features.7" in name or "features.8" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    model.classifier = nn.Sequential(
        nn.Linear(model.classifier[1].in_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model_2(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=EfficientNet_B0_Weights.IMAGENET1K_V1`. You can also use `weights=EfficientNet_B0_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 69.79%
Weighted Precision on the holdout dataset: 0.6984

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.69      0.72      0.71      5718
     Class 1       0.71      0.67      0.69      5703

    accuracy                           0.70     11421
   macro avg       0.70      0.70      0.70     11421
weighted avg       0.70      0.70      0.70     11421



### Extra Models

In [9]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_efficientnet_sgd_lr0.001.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_model_2(num_classes=2):
    model = models.efficientnet_b0(pretrained=True)

    for name, param in model.named_parameters():
        if "features.6" in name or "features.7" in name or "features.8" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    model.classifier = nn.Sequential(
        nn.Linear(model.classifier[1].in_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model_2(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=EfficientNet_B0_Weights.IMAGENET1K_V1`. You can also use `weights=EfficientNet_B0_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 76.23%
Weighted Precision on the holdout dataset: 0.7637

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.74      0.80      0.77      5718
     Class 1       0.78      0.73      0.75      5703

    accuracy                           0.76     11421
   macro avg       0.76      0.76      0.76     11421
weighted avg       0.76      0.76      0.76     11421



In [10]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

best_model_path = "/teamspace/studios/victorious-coral-h1k6/best_efficientnet_adam_lr0.001.pt"
new_dataset_image_path = "/teamspace/studios/written-violet-bdv7/holdout2_images.pt"
new_dataset_label_path = "/teamspace/studios/written-violet-bdv7/holdout2_labels.pt"

def get_model_2(num_classes=2):
    model = models.efficientnet_b0(pretrained=True)

    for name, param in model.named_parameters():
        if "features.6" in name or "features.7" in name or "features.8" in name:
            param.requires_grad = True
        else:
            param.requires_grad = False

    model.classifier = nn.Sequential(
        nn.Linear(model.classifier[1].in_features, 512),
        nn.BatchNorm1d(512),
        nn.ReLU(),
        nn.Dropout(0.5),
        nn.Linear(512, 256),
        nn.BatchNorm1d(256),
        nn.ReLU(),
        nn.Dropout(0.4),
        nn.Linear(256, 128),
        nn.BatchNorm1d(128),
        nn.ReLU(),
        nn.Dropout(0.3),
        nn.Linear(128, num_classes),
    )
    return model

model = get_model_2(num_classes=2)
model.load_state_dict(torch.load(best_model_path))
model = model.to(device)
model.eval()  

images = torch.load(new_dataset_image_path).to(device)
labels = torch.load(new_dataset_label_path).to(device)

batch_size = 32
holdout_dataset = TensorDataset(images, labels)
holdout_loader = DataLoader(holdout_dataset, batch_size=batch_size, shuffle=False)

correct = 0
total = 0
all_labels = []
all_predictions = []

with torch.no_grad():
    for images_batch, labels_batch in holdout_loader:
        outputs = model(images_batch)
        _, predicted = torch.max(outputs, 1)
        total += labels_batch.size(0)
        correct += (predicted == labels_batch).sum().item()
        all_labels.extend(labels_batch.cpu().tolist())
        all_predictions.extend(predicted.cpu().tolist())

accuracy = correct / total * 100

weighted_precision = precision_score(all_labels, all_predictions, average='weighted')

print(f"Accuracy on the holdout dataset: {accuracy:.2f}%")
print(f"Weighted Precision on the holdout dataset: {weighted_precision:.4f}")

print("\nClassification Report:")
print(classification_report(all_labels, all_predictions, target_names=["Class 0", "Class 1"]))

/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/zeus/miniconda3/envs/cloudspace/lib/python3.10/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=EfficientNet_B0_Weights.IMAGENET1K_V1`. You can also use `weights=EfficientNet_B0_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Accuracy on the holdout dataset: 74.62%
Weighted Precision on the holdout dataset: 0.7475

Classification Report:
              precision    recall  f1-score   support

     Class 0       0.73      0.78      0.76      5718
     Class 1       0.77      0.71      0.74      5703

    accuracy                           0.75     11421
   macro avg       0.75      0.75      0.75     11421
weighted avg       0.75      0.75      0.75     11421



| Model                         | Accuracy (%) | Weighted Precision |
|-------------------------------|--------------|--------------------|
| Base ResNet                   | 71.66        | 0.7190             |
| Base NASNet                   | 72.31        | 0.7232             |
| Base EfficientNet             | 58.67        | 0.6095             |
| Pretrained ResNet             | 57.95        | 0.5893             |
| Pretrained NASNet             | 71.38        | 0.7196             |
| Pretrained EfficientNet       | 56.78        | 0.5745             |
| ResNet (lr=0.01, Adam)        | 64.35        | 0.6449             |
| NASNet (lr=0.01, Adam)        | 66.72        | 0.6905             |
| EfficientNet (lr=0.01, Adam)  | 64.35        | 0.6476             |
| ResNet (lr=0.1, Adam)         | 52.53        | 0.6014             |
| NASNet (lr=0.1, Adam)         | 49.93        | 0.2493             |
| EfficientNet (lr=0.1, Adam)   | 50.07        | 0.2507             |
| ResNet (lr=0.2, Adam)         | 53.47        | 0.5876             |
| NASNet (lr=0.2, Adam)         | 49.93        | 0.2493             |
| EfficientNet (lr=0.2, Adam)   | 53.80        | 0.5397             |
| ResNet (lr=0.01, SGD)         | 73.53        | 0.7353             |
| NASNet (lr=0.01, SGD)         | 69.36        | 0.7188             |
| EfficientNet (lr=0.01, SGD)   | 77.16        | 0.7723             |
| ResNet (lr=0.1, SGD)          | 70.26        | 0.7031             |
| NASNet (lr=0.1, SGD)          | 70.22        | 0.7144             |
| EfficientNet (lr=0.1, SGD)    | 72.31        | 0.7231             |
| ResNet (lr=0.2, SGD)          | 66.93        | 0.6697             |
| NASNet (lr=0.2, SGD)          | 68.71        | 0.7005             |
| EfficientNet (lr=0.2, SGD)    | 69.79        | 0.6984             |
| ResNet18 + KNN                | 73.08        | 0.7318             |
| ResNet18 + SVM                | 73.44        | 0.7347             |
| EfficientNet + KNN            | 75.29        | 0.7529             |
| EfficientNet + SVM            | 75.33        | 0.7533             |
| NasNetLarge + KNN             | 72.72        | 0.7272             |
| NasNetLarge + SVM             | 72.84        | 0.7284             |
